### PR TITLE
Probability grid protos

### PR DESCRIPTION
**Pull request by macmason**
A first stab at being able to serialize for lifelong mapping.

### CODE REVIEW

**Comment by SirVer** [comment]
yip-iii!!! mac is back!

**Comment by macmason** [comment]
My clang-format munged some nearby-but-not-related code; let me know if you'd like me to un-munge.

File: cartographer/mapping_2d/map_limits_test.cc [hunk 1]
```diff
@@ -22,6 +22,32 @@ namespace cartographer {
...
+  EXPECT_EQ(limits.resolution(), proto.resolution());
+  EXPECT_EQ(limits.max().x(), proto.max().x());
+  EXPECT_EQ(limits.max().y(), proto.max().y());
+  EXPECT_EQ(ToProto(limits.cell_limits()).DebugString(),
+            proto.cell_limits().DebugString());
+}
+
+TEST(MapLimitsTest, ProtoConstructor) {
+  proto::MapLimits limits;
+  limits.set_resolution(1.0);
```
Comment (damonkohler) [inline cartographer/mapping_2d/map_limits_test.cc]: Consistency: drop the trailing 0s

File: cartographer/mapping_2d/probability_grid.h [hunk 2]
```diff
@@ -48,6 +49,23 @@ class ProbabilityGrid {
...
+        cells_(),
+        update_indices_(proto.update_indices().begin(),
+                        proto.update_indices().end()),
+        max_x_(proto.max_x()),
+        max_y_(proto.max_y()),
+        min_x_(proto.min_x()),
+        min_y_(proto.min_y()) {
+    cells_.reserve(proto.cells_size());
+    for (auto cell : proto.cells()) {
+      // TODO(macmason): DCHECK?
```
Comment (damonkohler) [inline cartographer/mapping_2d/probability_grid.h]: No DCHECKs, please. This is only for deserialization. It seems like performance shouldn't be a concern?

File: cartographer/mapping_2d/probability_grid.h [hunk 3]
```diff
@@ -154,6 +173,24 @@ class ProbabilityGrid {
     }
   }
 
+  proto::ProbabilityGrid ToProto() {
+    proto::ProbabilityGrid result;
+    *result.mutable_limits() = cartographer::mapping_2d::ToProto(limits_);
+    result.mutable_cells()->Reserve(cells_.size());
+    for (auto cell : cells_) {
```
Comment (damonkohler) [inline cartographer/mapping_2d/probability_grid.h]: const

File: cartographer/mapping_2d/probability_grid.h [hunk 4]
```diff
@@ -154,6 +173,24 @@ class ProbabilityGrid {
...
+      result.mutable_cells()->Add(cell);
+    }
+    result.mutable_update_indices()->Reserve(update_indices_.size());
+    for (auto update : update_indices_) {
```
Comment (damonkohler) [inline cartographer/mapping_2d/probability_grid.h]: const

File: cartographer/mapping_2d/probability_grid.h [hunk 5]
```diff
@@ -48,6 +49,23 @@ class ProbabilityGrid {
...
+      : limits_(proto.limits()),
+        cells_(),
+        update_indices_(proto.update_indices().begin(),
+                        proto.update_indices().end()),
+        max_x_(proto.max_x()),
+        max_y_(proto.max_y()),
+        min_x_(proto.min_x()),
+        min_y_(proto.min_y()) {
+    cells_.reserve(proto.cells_size());
+    for (auto cell : proto.cells()) {
```
Comment (damonkohler) [inline cartographer/mapping_2d/probability_grid.h]: const

File: cartographer/mapping_2d/probability_grid_test.cc [hunk 6]
```diff
@@ -17,13 +17,48 @@
...
 
 #include "gtest/gtest.h"
 
 namespace cartographer {
 namespace mapping_2d {
 namespace {
 
+TEST(ProbabilityGridTest, ProtoConstructor) {
+  proto::ProbabilityGrid proto;
+  const MapLimits limits(1.0, {2.0, 3.0}, CellLimits(4.0, 5.0));
```
Comment (damonkohler) [inline cartographer/mapping_2d/probability_grid_test.cc]: Trailing 0s

File: cartographer/mapping_2d/probability_grid_test.cc [hunk 7]
```diff
@@ -17,13 +17,48 @@
...
+TEST(ProbabilityGridTest, ToProto) {
+  ProbabilityGrid probability_grid(
+      MapLimits(1., Eigen::Vector2d(1., 1.), CellLimits(2, 2)));
+
+  const auto proto = probability_grid.ToProto();
+  EXPECT_EQ(ToProto(probability_grid.limits()).DebugString(),
+            proto.limits().DebugString());
+
+  // TODO(macmason): Figure out how to test the contents of cells_,
+  // update_indices_, and {min, max}_{x, y}_ gracefully.
```
Comment (damonkohler) [inline cartographer/mapping_2d/probability_grid_test.cc]: Probably some golden data tests are the most straight forward.

File: cartographer/transform/transform.h [hunk 8]
```diff
@@ -120,11 +120,13 @@ Eigen::Vector2d ToEigen(const proto::Vector2d& vector);
 Eigen::Vector3f ToEigen(const proto::Vector3f& vector);
 Eigen::Vector3d ToEigen(const proto::Vector3d& vector);
 Eigen::Quaterniond ToEigen(const proto::Quaterniond& quaternion);
+
```
Comment (damonkohler) [inline cartographer/transform/transform.h]: Remove

File: cartographer/mapping_2d/map_limits_test.cc (hunk 1 above)
Comment (macmason) [inline cartographer/mapping_2d/map_limits_test.cc]: Done.

File: cartographer/mapping_2d/probability_grid.h (hunk 5 above)
Comment (macmason) [inline cartographer/mapping_2d/probability_grid.h]: Done.

File: cartographer/mapping_2d/probability_grid.h (hunk 2 above)
Comment (macmason) [inline cartographer/mapping_2d/probability_grid.h]: Yeah, not too expensive. TODO removed, standard CHECK remains in place.

File: cartographer/mapping_2d/probability_grid_test.cc (hunk 6 above)
Comment (macmason) [inline cartographer/mapping_2d/probability_grid_test.cc]: Done.

File: cartographer/transform/transform.h (hunk 8 above)
Comment (macmason) [inline cartographer/transform/transform.h]: Done.

File: cartographer/mapping_2d/probability_grid_test.cc (hunk 7 above)
Comment (macmason) [inline cartographer/mapping_2d/probability_grid_test.cc]: I've left this here for the moment; mind if I create an issue and follow up?

File: cartographer/mapping_2d/probability_grid.h (hunk 3 above)
Comment (macmason) [inline cartographer/mapping_2d/probability_grid.h]: Done.

File: cartographer/mapping_2d/probability_grid.h (hunk 4 above)
Comment (macmason) [inline cartographer/mapping_2d/probability_grid.h]: Done.

File: cartographer/mapping_2d/probability_grid_test.cc (hunk 7 above)
Comment (damonkohler) [inline cartographer/mapping_2d/probability_grid_test.cc]: SGTM

**Pull request merged**